drivers/hmis/lib/ce_workflows/shared/ce_builder_utils.rb

```
@@ -0,0 +1,183 @@
+      # This populates the pools by calling the match engine with the same logic used in production
+      hit_max_iterations = false
+      10.times do
+        break unless Hmis::Ce::ChangeMarker.dirty.exists?
+
+        Hmis::Ce::ProcessChangesJob.new.perform(progress: progress)
+        hit_max_iterations = Hmis::Ce::ChangeMarker.dirty.exists?
+      end
+
+      return unless hit_max_iterations
+
+      Rails.logger.warn('CeBuilderUtils#build_candidate_pools reached maximum iterations (10). Dirty markers may not be fully processed.')
+    end
+
+    # Run this to keep state machine statuses in sync with custom statuses
+    def self.create_state_machine_custom_statuses(data_source)
+      Hmis::Ce::Referral.state_machine_states.map do |state|
+        status = Hmis::Ce::CustomReferralStatus.find_or_initialize_by(
+          key: state.to_s,
+          data_source: data_source,
+        )
+        label = case state.to_s
+        when 'rejected' then 'Declined'
+        else state.to_s.humanize.titleize
+        end
+        status.name = label
+        status.save!
+      end
+    end
+
+    def self.delete_template_and_associated_data(template_identifier)
+      raise 'This method destroys data and should not be run in production' if Rails.env.production?
+
+      puts "Deleting existing CE data associated with #{template_identifier}"
+
+      templates = Hmis::WorkflowDefinition::Template.where(identifier: template_identifier)
+      opportunities = Hmis::Ce::Opportunity.where(workflow_template_identifier: template_identifier)
+      instances = Hmis::WorkflowExecution::Instance.where(template: templates)
+      steps = Hmis::WorkflowExecution::Step.where(instance: instances)
+      referrals = Hmis::Ce::Referral.where(workflow_instance: instances)
+
+      Hmis::Ce::ReferralNote.where(referral: referrals).destroy_all
+      Hmis::Ce::ReferralParticipant.where(referral: referrals).destroy_all
+      referrals.destroy_all
```

this is probably fine but destroy_all may fail silently. It may be safer to use something like `find_each(&:destroy!)`